packages/fuels-accounts/src/provider.rs

```diff
@@ -521,49 +536,41 @@ impl Provider {
         })
     }
 
-    pub async fn estimate_transaction_cost<T: Transaction + Clone>(
+    pub async fn estimate_transaction_cost<T: Transaction>(
         &self,
-        tx: &T,
+        tx: T,
         tolerance: Option<f64>,
     ) -> Result<TransactionCost> {
         let NodeInfo { min_gas_price, .. } = self.node_info().await?;
-
+        let gas_price = std::cmp::max(tx.gas_price(), min_gas_price);
         let tolerance = tolerance.unwrap_or(DEFAULT_GAS_ESTIMATION_TOLERANCE);
-        let dry_run_tx = Self::generate_dry_run_tx(tx);
-        let consensus_parameters = self.consensus_parameters();
+
+        // Remove limits from an existing Transaction for accurate gas estimation
+        let new_tx = tx.to_dry_run_tx(0, MAX_GAS_PER_TX);
         let gas_used = self
-            .get_gas_used_with_tolerance(&dry_run_tx, tolerance)
+            .get_gas_used_with_tolerance(new_tx.clone(), tolerance)
             .await?;
-        let gas_price = std::cmp::max(tx.gas_price(), min_gas_price);
 
-        // Update the dry_run_tx with estimated gas_used and correct gas price to calculate the total_fee
-        dry_run_tx
-            .with_gas_price(gas_price)
-            .with_gas_limit(gas_used);
+        // Update the tx with estimated gas_used and correct gas price to calculate the total_fee
+        let new_tx = new_tx.to_dry_run_tx(gas_price, gas_used);
```

Comment:
IMO `to_dry_run_tx` feels a bit obscure to be a user facing method. It just set's two fields at once, while the name implies there is a higher purpose to it. The previous `generate_dry_run_tx` applied specific values to these fields so the name made more sense.
What do you think about just using two setters or going back to the old implementation?

Reply:
Thanks for the comment! I went back to the old implementation.